src/Geometry/K3Auto.jl

```diff
@@ -0,0 +1,2059 @@
+  # clear denominators
+  Lz = fmpz_mat[change_base_ring(ZZ,i*denominator(i)) for i in Lq]
+  # primitive in S
+  Lz = fmpz_mat[divexact(i,gcd(vec(i))) for i in Lz]
+  @hassert :K3Auto 2 all(all(x>=0 for x in vec(r*gram_matrix(D.data.SS)*transpose(i))) for i in Lz)
+  return Lz
+end
+
+function Base.show(io::IO, c::K3Chamber)
+  if isdefined(c,:walls)
+    print(IOContext(io, :compact => true), "Chamber  in dimension $(length(walls(c)[1])) with $(length(walls(c))) walls")
+  else
+    print(IOContext(io, :compact => true), "Chamber: $(c.weyl_vector[1,1:rank(c.data.S)])")
+  end
+end
+
+@doc Markdown.doc"""
+    fingerprint(D::K3Chamber)
+
+Return the fingerprint of this chamber.
+
+The fingerprint is an invariant computed from the rays and their inner products.
+"""
+function fingerprint(D::K3Chamber)
+  v = sum(walls(D))
+  G = D.data.gramS
+  m1 = (v*G*transpose(v))[1,1]
+  m2 = [(a*G*transpose(a))[1,1] for a in walls(D)]
+  sort!(m2)
+  m3 = [(v*G*transpose(a))[1,1] for a in walls(D)]
+  sort!(m3)
+  m4 = fmpz[]
+  for i in 1:length(walls(D))
+    for j in 1:i-1
+      push!(m4,(walls(D)[i]*G*transpose(walls(D)[j]))[1,1])
+    end
+  end
+  sort!(m4)
+  V = Dict{Tuple{fmpz,fmpz},Vector{fmpz_mat}}()
+  for w in walls(D)
+    i =  (v*G*transpose(w))[1,1]
+    j =  (w*G*transpose(w))[1,1]
+    if (i,j) in keys(V)
+      push!(V[(i,j)],w)
+    else
+      V[(i,j)] = [w]
+    end
+  end
+  #=
+  # so far m5 was not needed to separate the O(S)-orbits
+  m5 = []
+  for i in keys(V)
+    vi = sum(V[i])
+    push!(m5, [i,sort!([(vi*G*transpose(j))[1,1] for j in walls(D)])])
+  end
+  sort!(m5)
+  # So far we have only O(S)-invariants. There are also ways to produce G-invariants
+  # by working the the images of the rays in the discriminant group and their
+  # G-orbits. Perhaps one has to switch to S^\vee primitive vectors in this case.
+  =#
+  return (m1, m2, m3, m4)
+end
+
+
+"""
+Compute the fingerprint of Plesken-Souvignier and change the basis
+matrix and gram matrix accordingly.
+
+It is computed from the walls and the gram matrix of `S`.
+
+a permutation per for the
+order of the basis-vectors is chosen
+such that in every step the number of
+possible continuations is minimal,
+for j from per[i] to per[dim-1] the
+value f[i][j] in the fingerprint f is
+the number of vectors, which have the
+same scalar product with the
+basis-vectors per[0]...per[i-1] as the
+basis-vector j and the same length as
+this vector with respect to all
+invariant forms
+"""
+function _fingerprint_backtrack!(D::K3Chamber)
+  n = rank(D.data.S)
+  V = walls(D)
+  gramS = gram_matrix(D.data.S)
+  B, indB = _find_basis(V, n)
+  tmp = V[indB]
+  deleteat!(V, indB)
+  prepend!(V, tmp)
+  lengths = fmpq[(v*gramS*transpose(v))[1,1] for v in V]
+  D.lengths = lengths
+  gramB = change_base_ring(ZZ, B*gramS*transpose(B))
+  D.gramB = gramB
+  D.B = B
+
+  per = Vector{Int}(undef, n)
+  for i in 1:n
+    per[i] = i
+  end
```

Comment:
```suggestion
  per = collect(1:n)
```

Reply:
Does this differ in performance or in style only?